### ISSUE TITLE
Decryption (fde) don't work 

### ISSUE THREAD

**Issue by bassbounce** [TeamWin/Team-Win-Recovery-Project#1322]
Hello, 

I have compiled the latest twrp for pocophone, but it cannot decrypt data when booting stock miui 10 rom, do anyone have fix this for other Xiaomi phones? Since the are using fde encryption.

**Comment by CaptainThrowback** [comment]
Decryption issues need to be fixed by the device maintainer. There's nothing the TWRP developers can do about it. Please contact your device maintainer, or go to #twrp on Freenode and try to get some specific assistance for adding decryption to the device tree there.

**Comment by DavidBerdik** [comment]
With all due respect, are you sure of this? The general consensus seems to be that it's a TWRP problem given the number of users that are affected by it. (My phone: Nexus 6/shamu running LineageOS 15.1.)

**Comment by CaptainThrowback** [comment]
@DavidBerdik yes I'm positive. All of the devices I maintain use FDE and they all work fine, because I've updated the device trees to make sure that they do. The TWRP code functions perfectly fine for FDE, but the device trees have to be updated to make it work. This is a device maintainer issue.

**Comment by DavidBerdik** [comment]
@CaptainThrowback For the purposes of enabling me to provide more useful information to my device maintainer, what about the device tree needs to be updated in order for it to work properly?

**Comment by CaptainThrowback** [comment]
That really depends on the device and Android version. Which is why I suggested sending them to IRC to get direct help from the TWRP team. Decryption can get tricky. It's not exactly the same on any of my devices, though it is similar. Looking at the device trees I maintain might be helpful, but HTC devices aren't necessarily the same as other devices. There are also multiple ways to make it work. So unfortunately there's no easy answer to that question.

**Comment by DavidBerdik** [comment]
Well in that case, I will open an issue with my maintainer and point them here. Unfortunately it's not much help for them, but at least it will demonstrate to them that it's a device problem and not a TWRP problem.

**Comment by cosmicdan** [comment]
I've looked into this a lot. It bugs me because, while Poco can be simply modded for FBE working; TWRP seems to lack support for restoring userdata backup (nandroid) that were made under FBE. Getting basic FDE working in device tree first is ideal since we know that TWRP should support it 100%.

Here's what I've learned about this so far:

 - There is no keystore module in the TWRP device tree for other recent Xiaomi devices. My tree here also reflects this: https://github.com/CosmicDan-Android/twrp_device_xiaomi_beryllium
 - Mounting /system (a tip I discovered from a Pixel thread) allows it to find the keystore module at lib/hw/ (there may be other fixes I had to make but I can't remember if they were specific to FBE) however it seems to be a generic "OpenSSL Keystore", which is not something I've seen in TWRP log before - seems to be more of a template keystore from AOSP or something
 - At this point, trying to decrypt results in logcat error that it cannot decrypt because it was encrypted with a different keystore, or something like that (googling the error returned absolutely zero results, IIRC).

This suggests that the standard ROM in Poco is in-fact not using that keystore "module", which is probably just software encryption or a generic stub or something. Although if that's true it begs the question - why would that module exist in system at all?

I've tried looking for some other keystore module, even resorted to searching string dumps of all libs, but couldn't find anything.

Regardless, I am 99% sure that this is a device tree issue - not a TWRP one - and we just need someone more experienced to take a look at it. I saw Dees_Troy was one of the recipients of the XDA dev stimulus program, so hopefully it won't be long :)

**Comment by CaptainThrowback** [comment]
It would be nice if @Dees-Troy was able to update the FDE code in TWRP to work on later Android versions. Maybe the new device he's receiving will help with that. Currently the devices I maintain are using vold decrypt for FDE decryption, which works, but was originally intended as a fallback solution if TWRP crypto failed. It's become more of a stop-gap solution than was previously intended, but it offers a lot of flexibility for multiple types of FDE encryption.

I guess we'll see how he fares with the new device.

**Comment by DavidBerdik** [comment]
I just opened an issue on the shamu device tree: https://github.com/TeamWin/android_device_moto_shamu/issues/1

When you mentioned earlier that it's a device tree issue, I misunderstood you to mean that it is a problem with the LineageOS device tree for the Nexus 6 and not the device tree for TWRP.

**Comment by DavidBerdik** [comment]
@CaptainThrowback As a followup to the issue that I opened on the device tree, I posted a log from TWRP. Sadly I have not heard anything back yet (I tried the IRC channel as well, but no luck). Now that my semester is over, I have some time to look in to fixing it myself which I am most certainly doing, but I am having a bit of an issue with handling the learning curve for building TWRP.

One trick that I thought up to solve the issue, and I am not sure if this is sensible or possible, is to try building TWRP using the Nexus 6 device tree maintained by the LineageOS team. I would think that by using that device tree instead, it would resolve the decryption issue, but again, I am not sure if that is possible or would even work.

**Comment by cosmicdan** [comment]
The original issue has been resolved at https://github.com/vasishath/android_device_xiaomi_beryllium/tree/android-8.1-vold - originally using vold decrypt but has since been changed to a custom cryptfs_hw source, but that could be staging work for MultiROM.

The last release of TWRP built against this tree may still use vold. You should be able to find what you need in the commit history regardless.
-- 
Sent with <a href="https://play.google.com/store/apps/details?id=com.onegravity.k10.pro2"><b>K-@ Mail</b></a> - the evolution of emailing.

**Comment by DavidBerdik** [comment]
@cosmicdan - Okay so I am looking through the commit history and I think I found the commit you are referring to. I am not certain, but as far as I can tell the other ones are irrelevant.

[https://github.com/vasishath/android_device_xiaomi_beryllium/commit/e9c3709979eb802bc42cd84215945adf11dbf160](https://github.com/vasishath/android_device_xiaomi_beryllium/commit/e9c3709979eb802bc42cd84215945adf11dbf160)

Even if this is the correct one, I still do not know what exactly I would have to change in order to get this to work. I am still trying to figure out how device trees work.

**Comment by Akipe** [comment]
@DavidBerdik  @cosmicdan Hi ! I think I managed to port twrp 3.3.1 on beryllium with decryption support (it works for me, but I need feedbacks). You can test my build here : https://forum.xda-developers.com/poco-f1/development/recovery-twrp-3-3-1-beta01-touch-t3987567
